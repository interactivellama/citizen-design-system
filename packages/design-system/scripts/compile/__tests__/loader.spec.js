// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

/* eslint-env jest */

import React from 'react';
import ReactDOM from 'react-dom/server';
import loader from '../mdx-post-loader';
import createHelper from '../mdx-helper';

describe('loader template', () => {
  it('snapshots the template', () => {
    expect(loader('export default function')).toMatchSnapshot();
  });
});

describe('mdx helper', () => {
  let helper;

  beforeEach(() => {
    helper = createHelper();
  });

  it('renders h2', () => {
    const createH2 = helper.HeadingEl('h2');
    const r = ReactDOM.renderToStaticMarkup(createH2({ id: 'hello' }, 'Hello'));
    expect(r).toEqual(
      ReactDOM.renderToStaticMarkup(
        <h2 id="hello" className="doc">
          Hello<a className="doc doc-anchor" href="#hello">
            #
          </a>
        </h2>
      )
    );
  });

  describe('after h2 rendered', () => {
    let createH2;
    beforeEach(() => {
      createH2 = helper.HeadingEl('h2');
      createH2({ id: 'accordion' }, 'Accordion');
    });

    it('added to toc', () => {
      expect(helper.toc).toEqual([
        { id: 'accordion', headingLevel: '2', title: 'Accordion', children: [] }
      ]);
    });

    it('adds subheadings as children', () => {
      const createH3 = helper.HeadingEl('h3');
      createH3({ id: 'states' }, 'States');
      const item = helper.toc[0];
      expect(item.id).toEqual('accordion');
      expect(item.children[0].id).toEqual('states');
    });

    it('adds subheadings as siblings', () => {
      const createH3 = helper.HeadingEl('h3');
      createH3({ id: 'states' }, 'States');
      createH3({ id: 'states2' }, 'States');
      const item = helper.toc[0];
      expect(item.id).toEqual('accordion');
      expect(item.children[0].id).toEqual('states');
      expect(item.children[1].id).toEqual('states2');
    });

    it('adds subheadings as childrens children', () => {
      const createH3 = helper.HeadingEl('h3');
      const createH4 = helper.HeadingEl('h4');
      createH3({ id: 'states' }, 'States');
      createH4({ id: 'states2' }, 'States');
      const item = helper.toc[0];
      expect(item.id).toEqual('accordion');
      expect(item.children[0].id).toEqual('states');
      expect(item.children[0].children[0].id).toEqual('states2');
    });

    describe('after child h3 has rendered', () => {
      beforeEach(() => {
        const createH3 = helper.HeadingEl('h3');
        createH3({ id: 'child1' }, 'child');
      });

      it('adds a new parent h2', () => {
        createH2({ id: 'chat' }, 'Chat');
        const lastItem = helper.toc[helper.toc.length - 1];
        expect(lastItem.id).toEqual('chat');
      });
    });
  });
});
