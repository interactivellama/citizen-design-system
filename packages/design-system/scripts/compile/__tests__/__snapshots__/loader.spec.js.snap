// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader template snapshots the template 1`] = `
"
  import Doc from '../../../shared/components/Doc';

  const addDocClass = className => ['doc', className].filter(x => x).join(' ')
  const last = xs => xs[xs.length - 1]
  const findChildHeadingByLevel = (level, heading = {}) =>
  heading.headingLevel < level ?
  findChildHeadingByLevel(level, last(heading.children)) || heading :
  null

  const El = el => (props, ...kids) => {
  const className = addDocClass(props.className);
  return React.createElement(
  el,
  Object.assign(props, { className, key: key() }),
  kids);

}
  const HeadingEl = el => (props, ...kids) => {
    const className = addDocClass(props.className);
    const [_, headingLevel] = el.split('h');

    const lastParentHeading =
    toc.length && findChildHeadingByLevel(headingLevel, last(toc));
    const item = {
      id: props.id,
      key: props.id,
      title: kids[0],
      headingLevel,
      children: [] };


    lastParentHeading ? lastParentHeading.children.push(item) : toc.push(item);

    return React.createElement(
    el,
    Object.assign(props, { className }),
    kids,
    React.createElement(
    'a',
    {
      className: 'doc doc-anchor',
      href: \`#\${props.id}\`,
      ['aria-label']: 'Jump to section' },

    '#'));


  }

  let counter = 0
  const key = () => {
  const id = ++counter;
  return id;
}

  const factories = {h1: HeadingEl('h1'),h2: HeadingEl('h2'),h3: HeadingEl('h3'),h4: HeadingEl('h4'),h5: HeadingEl('h5'),h6: HeadingEl('h6'), p: El('p'),div: El('div'),a: El('a'),em: El('em'),strong: El('strong'),ol: El('ol'),ul: El('ul'),li: El('li'),code: El('code'),blockquote: El('blockquote'),pre: El('pre'),tr: El('tr'),td: El('td'),th: El('th'),table: El('table'),thead: El('thead'),tbody: El('tbody')}

  let toc = [];
  export const getToc = () => toc;

    export const Content = function
    export default props => {
      toc = [];
      return <Doc><Content {...props} factories={factories} /></Doc>
    }
  "
`;
